### DISCUSSION TITLE
Cut segment based on frames not on time (keyframes)

### DISCUSSION THREAD

**Discussion by DGrv** [mifi/lossless-cut#1702] [Ideas]
Hi,

I just wrote a small lua script to used in mpv player to be able to cut at the exact time cursor.
After a lot of different test I am happy with the result based on frame cut. 
Seems to be exact cut frame, the audio seems fine and no duplicates frames.

What is your point of view about it and what do you think of a new cut function based on frames ?
I have however no idea of the reproducibility of this with other extension and codec (using mp4 and h264).

The idea is as follow:

- get the time where to cut $( ct )$
- get the video frame rate $( fr )$ :  `ffprobe -v error -select_streams v:0 -show_entries stream=r_frame_rate -print_format default=nokey=1:noprint_wrappers=1 input.mp4`
- calculate at which frame to cut for seg1 $( cvf1 = \lfloor ct*fr \rfloor)$
- calculate at which frame to cut for seg2 $( cvf2 = cvf1 + 1 )$
- get the number of video frame of the video $( vf )$ : `ffprobe -v error -select_streams v:0 -count_frames -show_entries stream=nb_read_frames -print_format default=nokey=1:noprint_wrappers=1 input.mp4 `
- get the number of audio frame of the audio $( af )$ : `ffprobe -v error -select_streams a:0 -count_frames -show_entries stream=nb_read_frames -print_format default=nokey=1:noprint_wrappers=1 input.mp4`
- calculate the audio cut frame $( caf2 = \lfloor \frac{af}{vf} * cvf2 \rfloor  )$
- cut seg1 : ` ffmpeg -i input.mp4 -frames:v cvf1 -c:v copy -c:a copy output_seg1.mp4 `
- cut seg2 : ` ffmpeg -i input.mp4 -vf select="between(n\, cf2 \, vf ),setpts=PTS-STARTPTS" -af aselect="between(n\, caf2 \, af ),asetpts=PTS-STARTPTS" output_seg2.mp4 `

Practical example is here :

**input.mp4:**
```sh
fr=29.97
ct=3.6703333333333
cvf1=109
vf=321
af=498
caf2=170
ffmpeg -i "input.mp4" -frames:v 109 -c:v copy -c:a copy -y "output_seg1.mp4" 
ffmpeg -i "input.mp4" -vf select="between(n\,110\,321),setpts=PTS-STARTPTS" -af aselect="between(n\,170\,498),asetpts=PTS-STARTPTS" -y "output_seg2.mp4" 
```

[Lua script is available here](https://github.com/DGrv/dorian.gravier.github.io/blob/master/files/Lua/cut_v04.lua)

**Comment by mifi** [comment]
Hi and thanks for your research




> * cut seg2 : `ffmpeg -i input.mp4 -vf select="between(n\, cf2 \, vf ),setpts=PTS-STARTPTS" -af aselect="between(n\, caf2 \, af ),asetpts=PTS-STARTPTS" output_seg2.mp4`

Isn't this a lossy operation that will re-encode the whole video?

**Reply by DGrv** [reply]
I do not know if it is lossy but it will for sure re-encode. I did not search for frame cut that is not re-encoding but seems not possible.

Meanwhile I found [trim.lua](https://github.com/aerobounce/trim.lua) or [mpv-plugin-excerpt](https://gitlab.com/lvml/mpv-plugin-excerpt) via [this post in stackoverflow](https://stackoverflow.com/questions/65522371/getting-time-codes-in-video-files-easier/65524847#65524847) that might be of interest that is doing this lossless but still using keyframe meaning not as precise as I would wish.

However seems that my wish to cut exactly is not possible without reencoding... sure I understand your point and my idea does not fit to losslesscut purpose then, I did not think so far and was obsessed by an exact cutting.
The combination of losslesscut and then my script is then the way to go for me regarding or I edit my video.

Thanks for the short and useful answer :)

**Reply by mifi** [reply]
Ok! Thanks for sharing anyways, hopefully useful to some. I think lossy cutting&encoding is out of the scope for LosslessCut as for now, but maybe some time in the future:)

**Comment by DGrv** [comment]
Sure.

I think I should explain a bit how I came to this frame cut idea.

Keyframe cut is just perfect when I wanna cut non adjacent segments. 
My problem is that I wanna cut adjacent segment from time to time and the shortcut "b" is great but the output segments are not optimal since if you merged them again you have duplicated frames at the cut position. This is a ffmpeg issue and I tried many things to avoid this. Until now the frame cut seems to be from far the best (except that you need to reencode).

One idea would be cut the segments as it is yes and use the first cmd that I proposed to remove the last or the 2 last frames of the seg1 (which would be also lossless). The only thing is that I am not sure if it is always 1 or 2 frames that are duplicated or even more maybe.
But this would be an option I guess to fix the duplicated frames if you merge the 2 segments.
Your point of view ?

**Reply by mifi** [reply]
haven't found a definite way to seek to an exact frame in ffmpeg that works with all files. see #1216
